presto-main/src/main/java/io/prestosql/execution/scheduler/SubnetBasedTopology.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.execution.scheduler;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Ordering;
+import io.prestosql.spi.HostAddress;
+
+import javax.inject.Inject;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.execution.scheduler.NetworkLocation.ROOT_LOCATION;
+import static java.util.Objects.requireNonNull;
+
+public class SubnetBasedTopology
+        implements NetworkTopology
+{
+    private final List<byte[]> subnetMasks;
+    private final AddressProtocol protocol;
+
+    @Inject
+    public SubnetBasedTopology(SubnetTopologyConfig config)
+    {
+        this(config.getCidrPrefixLengths(), config.getAddressProtocol());
+    }
+
+    SubnetBasedTopology(List<Integer> cidrPrefixLengths, AddressProtocol protocol)
+    {
+        requireNonNull(cidrPrefixLengths, "cidrPrefixLengths is null");
+        requireNonNull(protocol, "protocol is null");
+
+        validateHierarchy(cidrPrefixLengths, protocol);
+
+        this.protocol = protocol;
+        this.subnetMasks = cidrPrefixLengths.stream()
+                .map(protocol::computeSubnetMask)
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public NetworkLocation locate(HostAddress address)
+    {
+        try {
+            InetAddress inetAddress = protocol.getInetAddress(address.getAllInetAddresses());
+
+            if (inetAddress == null) {
+                return ROOT_LOCATION;
+            }
+
+            byte[] addressBytes = inetAddress.getAddress();
+            ImmutableList.Builder<String> segments = ImmutableList.builder();
+
+            for (byte[] subnetMask : subnetMasks) {
+                byte[] bytes = applyMask(addressBytes, subnetMask);
+                segments.add(InetAddress.getByAddress(bytes).getHostAddress());
+            }
+
+            segments.add(inetAddress.getHostAddress());
```

Comment:
You might want to use `InetAddresses.toAddrString()` here, so that you get correctly formatted IPv6 addresses.